### PR TITLE
Fixing #241

### LINKED ISSUE #241: Flow: identifier `ESLintContext`. Could not resolve name
Using Flow 0.42.0 and eslint-plugin-jsx-a11y 5.0.1.

Got the following error when running Flow check on my project:

```
$ flow check 

node_modules/eslint-plugin-jsx-a11y/src/rules/interactive-supports-focus.js:46
 46:   create: (context: ESLintContext) => ({
                         ^^^^^^^^^^^^^ identifier `ESLintContext`. Could not resolve name

node_modules/eslint-plugin-jsx-a11y/src/rules/media-has-caption.js:43
 43:   create: (context: ESLintContext) => ({
                         ^^^^^^^^^^^^^ identifier `ESLintContext`. Could not resolve name

node_modules/eslint-plugin-jsx-a11y/src/rules/no-interactive-element-to-noninteractive-role.js:47
 47:   create: (context: ESLintContext) => {
                         ^^^^^^^^^^^^^ identifier `ESLintContext`. Could not resolve name

node_modules/eslint-plugin-jsx-a11y/src/rules/no-interactive-element-to-noninteractive-role.js:51
 51:         attribute: ESLintJSXAttribute,
                        ^^^^^^^^^^^^^^^^^^ identifier `ESLintJSXAttribute`. Could not resolve name

node_modules/eslint-plugin-jsx-a11y/src/rules/no-noninteractive-element-interactions.js:46
 46:   create: (context: ESLintContext) => {
                         ^^^^^^^^^^^^^ identifier `ESLintContext`. Could not resolve name

node_modules/eslint-plugin-jsx-a11y/src/rules/no-noninteractive-element-to-interactive-role.js:47
 47:   create: (context: ESLintContext) => {
                         ^^^^^^^^^^^^^ identifier `ESLintContext`. Could not resolve name

node_modules/eslint-plugin-jsx-a11y/src/rules/no-noninteractive-element-to-interactive-role.js:51
 51:         attribute: ESLintJSXAttribute,
                        ^^^^^^^^^^^^^^^^^^ identifier `ESLintJSXAttribute`. Could not resolve name

node_modules/eslint-plugin-jsx-a11y/src/rules/no-noninteractive-tabindex.js:45
 45:   create: (context: ESLintContext) => {
                         ^^^^^^^^^^^^^ identifier `ESLintContext`. Could not resolve name

node_modules/eslint-plugin-jsx-a11y/src/rules/no-noninteractive-tabindex.js:49
 49:           node: JSXOpeningElement,
                     ^^^^^^^^^^^^^^^^^ identifier `JSXOpeningElement`. Could not resolve name

node_modules/eslint-plugin-jsx-a11y/src/rules/no-static-element-interactions.js:47
 47:   create: (context: ESLintContext) => {
                         ^^^^^^^^^^^^^ identifier `ESLintContext`. Could not resolve name


Found 10 errors
error Command failed with exit code 2.
```

Looks like `src/` folder should be excluded from the module.

### PR DESCRIPTION
**Pull request by leos**
Making the flow type imports explicit. This also revealed a potential issue in no-noninteractive-element-to-interactive-role.js.

Fixes #241.

### CODE REVIEW

**Comment by ljharb** [comment]
Seems legit to me, but I'll defer to reviewers who are more familiar with Flow.

File: src/rules/no-noninteractive-element-to-interactive-role.js [hunk 1]
```diff
@@ -50,7 +53,7 @@ module.exports = {
       JSXAttribute: (
         attribute: ESLintJSXAttribute,
       ) => {
-        const attributeName: JSXIdentifier = attribute.name.name;
+        const attributeName: JSXIdentifier = propName(attribute);
```
Comment (ljharb) [inline src/rules/no-noninteractive-element-to-interactive-role.js]: it would be good to add a regression test on this tho, if this makes a difference.

File: flow/eslint.js [hunk 2]
```diff
@@ -1,9 +1,12 @@
...
+/*
+ * @flow
+ */
+export type ESLintReport = {
   node: any,
   message: string,
 };
 
-type ESLintContext = {
+export type ESLintContext = {
```
Comment (evcohen) [inline flow/eslint.js]: isn't there a way to make types global for your project?

Comment (leos) [inline flow/eslint.js]: I'm not an expert on Flow, but it can be done is a [libdef](https://flow.org/en/docs/libdefs/creation/#toc-declaring-a-global-type). The way it is currently being done is problematic for two reasons:

1) It requires an entry in .flowconfig which doesn't work correctly when the project is sitting in  node_modules (see #241)
2) Again, not sure what the underlying reason was but it looks like doing it that way masked a bug - `attribute.name.name`.

There's a bigger decision which you guys as project owners should make is whether you want to make these types of imports explicit. Coming from Python, IMO explicit is better than implicit.

Comment (ljharb) [inline flow/eslint.js]: fwiw, I'm a huge fan of explicit over implicit, and I've never written Python. We should be trying to make fewer globals, not more.

File: src/rules/no-noninteractive-element-to-interactive-role.js (hunk 1 above)
Comment (jessebeach) [inline src/rules/no-noninteractive-element-to-interactive-role.js]: Doesn't make a difference and in fact I should have used the `propName` function originally to deal with namespaces.

File: flow/eslint.js (hunk 2 above)
Comment (jessebeach) [inline flow/eslint.js]: Well that's just really weird. I reproduced the typing issue raised in `no-noninteractive-element-to-interactive-role` when the eslint types are explicitly exported. Nice catch @leos .

I agree with @ljharb here on explicit vs. implicit.

File: src/rules/no-noninteractive-element-to-interactive-role.js (hunk 1 above)
Comment (ljharb) [inline src/rules/no-noninteractive-element-to-interactive-role.js]: If it doesn't make a difference tho, why does the `propName` function need to exist?

If `propName` handles some edge case, then it'd be good to have a regression test that fails without usage of `propName`.

File: flow/eslint.js (hunk 2 above)
Comment (evcohen) [inline flow/eslint.js]: cool! agreed, just wondering 😄

File: src/rules/no-noninteractive-element-to-interactive-role.js (hunk 1 above)
Comment (evcohen) [inline src/rules/no-noninteractive-element-to-interactive-role.js]: it does handle multiple cases- an attribute can have a namespace:

https://github.com/evcohen/jsx-ast-utils/blob/master/src/propName.js#L9

Comment (ljharb) [inline src/rules/no-noninteractive-element-to-interactive-role.js]: Great - let's add a regression test with a namespaced attribute then :-)

**Comment by leos** [comment]
Sorry, just to clarify, do you want me to make the test update as part of the PR?

**Comment by evcohen** [comment]
> Sorry, just to clarify, do you want me to make the test update as part of the PR?

@leos yes please, since we are updating to `propNames` in this PR, we should be adding the test in same PR. thanks for all of your work with this!

**Comment by leos** [comment]
Ok, can one of you either point me to a similar test that I can replicate or lay out what the logic I'm testing for is? I copied the `propName()` usage from `no-interactive-element-to-noninteractive-role.js` just to clear the Flow error. Briefly glancing though the tests I could find on that tests namespaces (and there are several other uses of `propName()`).

I don't grok what this module does or what interactive/noninteractive roles are - should `<input type="img" mycustomnamespace:role="img" />` be valid? If so, the code in the PR won't act on it since `propName` will return `mycustomnamespace:role`. I assume the original code would have had the same behavior.

Might be my unfamiliarity with the test framework but trying to add something like this to `alwaysValid` in the tests :
```
  /* Namespaced roles are not checked */
  { code: '<div mynamespace:role="term">' },
  { code: '<input mynamespace:role="img">' },
```
results in a `Parsing error: Unexpected token`.

**Comment by ljharb** [comment]
You seem to be missing a closing slash; not sure if that's the problem.

**Comment by leos** [comment]
Thanks @ljharb ! That fixed it, is the intent/logic correct?

**Comment by evcohen** [comment]
@leos this LGTM - @jessebeach @ljharb does this require a patch release?

**Comment by evcohen** [comment]
[v5.0.2](https://github.com/evcohen/eslint-plugin-jsx-a11y/releases/tag/5.0.2) published with fix. Thanks for all of your help @leos & everybody!!! let me know if its resolved in `v5.0.2`

**Comment by leos** [comment]
Ugh, I'm sorry, when it's packaged up apparently the 'flow' directory doesn't come along for the ride so the imports are broken.
```
node_modules/eslint-plugin-jsx-a11y/src/rules/interactive-supports-focus.js:16
 16: import type { ESLintContext } from '../../flow/eslint';
                                        ^^^^^^^^^^^^^^^^^^^ ../../flow/eslint. Required module not found

node_modules/eslint-plugin-jsx-a11y/src/rules/media-has-caption.js:13
 13: import type { ESLintContext } from '../../flow/eslint';
                                        ^^^^^^^^^^^^^^^^^^^ ../../flow/eslint. Required module not found

node_modules/eslint-plugin-jsx-a11y/src/rules/no-interactive-element-to-noninteractive-role.js:23
 23: import type { ESLintContext } from '../../flow/eslint';
                                        ^^^^^^^^^^^^^^^^^^^ ../../flow/eslint. Required module not found

node_modules/eslint-plugin-jsx-a11y/src/rules/no-interactive-element-to-noninteractive-role.js:24
 24: import type { ESLintJSXAttribute } from '../../flow/eslint-jsx';
                                             ^^^^^^^^^^^^^^^^^^^^^^^ ../../flow/eslint-jsx. Required module not found

node_modules/eslint-plugin-jsx-a11y/src/rules/no-noninteractive-element-interactions.js:22
 22: import type { ESLintContext } from '../../flow/eslint';
                                        ^^^^^^^^^^^^^^^^^^^ ../../flow/eslint. Required module not found

node_modules/eslint-plugin-jsx-a11y/src/rules/no-noninteractive-element-to-interactive-role.js:25
 25: import type { ESLintContext } from '../../flow/eslint';
                                        ^^^^^^^^^^^^^^^^^^^ ../../flow/eslint. Required module not found

node_modules/eslint-plugin-jsx-a11y/src/rules/no-noninteractive-element-to-interactive-role.js:26
 26: import type { ESLintJSXAttribute } from '../../flow/eslint-jsx';
                                             ^^^^^^^^^^^^^^^^^^^^^^^ ../../flow/eslint-jsx. Required module not found

node_modules/eslint-plugin-jsx-a11y/src/rules/no-noninteractive-tabindex.js:23
 23: import type { ESLintContext } from '../../flow/eslint';
                                        ^^^^^^^^^^^^^^^^^^^ ../../flow/eslint. Required module not found

node_modules/eslint-plugin-jsx-a11y/src/rules/no-static-element-interactions.js:23
 23: import type { ESLintContext } from '../../flow/eslint';
                                        ^^^^^^^^^^^^^^^^^^^ ../../flow/eslint. Required module not found
```

I think the best approach is to move flow into `src` with everything else.

**Comment by evcohen** [comment]
I think we can just remove `flow` from `.npmignore`

**Comment by leos** [comment]
That would work too. Do you want me to do the PR?
